proposals/paddle-operator-proposal.md

```diff
@@ -0,0 +1,236 @@
+    paddle-res-name: wide-ande-deep-service-ps-0
+  sessionAffinity: None
+  type: ClusterIP
+```
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wide-ande-deep-ps-0
+  namespace: paddle-system
+  ownerReferences:
+  - apiVersion: batch.paddlepaddle.org/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: PaddleJob
+    name: wide-ande-deep
+    uid: f206587f-5dee-46f5-9399-e835bde02487
+  resourceVersion: "9506900"
+  selfLink: /api/v1/namespaces/paddle-system/pods/wide-ande-deep-ps-0
+  uid: 36b27c8f-9712-474b-b21b-dd6b54aaef29
+spec:
+  containers:
+  - env:
+    - name: POD_IP
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: status.podIP
+    - name: PADDLE_TRAINER_ID
+      value: "0"
+    - name: TRAINING_ROLE
+      value: PSERVER
+    - name: PADDLE_TRAINING_ROLE
+      value: PSERVER
+    envFrom:
+    - configMapRef:
+        name: wide-ande-deep
+    image: registry.baidubce.com/paddle-operator/demo-wide-and-deep:v1
+```
+
+### Resulting Worker
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wide-ande-deep-worker-0
+  namespace: paddle-system
+  ownerReferences:
+  - apiVersion: batch.paddlepaddle.org/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: PaddleJob
+    name: wide-ande-deep
+    uid: f206587f-5dee-46f5-9399-e835bde02487
+  resourceVersion: "9507629"
+  selfLink: /api/v1/namespaces/paddle-system/pods/wide-ande-deep-worker-0
+  uid: e8534fe6-7c2e-4849-9a99-ffdcd5df76bb
+spec:
+  containers:
+  - env:
+    - name: POD_IP
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: status.podIP
+    - name: PADDLE_TRAINER_ID
+      value: "0"
+    - name: TRAINING_ROLE
+      value: TRAINER
+    - name: PADDLE_TRAINING_ROLE
+      value: TRAINER
+    envFrom:
+    - configMapRef:
+        name: wide-ande-deep
+    image: registry.baidubce.com/paddle-operator/demo-wide-and-deep:v1
+```
+
+The worker spec generates a pod. Currently worker will communicate to the master through the master's service name, we'll use a service registry for service discovery.
+
+## Design
+
+Here are some original design docs for paddle-perator on Kubernetes.
+ 
+* Paddle Operator Architecture on Kubernetes, please check out [design-arch](https://github.com/PaddleFlow/paddle-operator/blob/main/docs/design-arch.md)  
+* Paddle training job instance fault tolerant, please check out [design-fault-tolerant](https://github.com/PaddleFlow/paddle-operator/blob/main/docs/design-fault-tolerant.md)
+* Co-scheduling training job to prevent job instances from resource deadlock, please check out [design-coschedule](https://github.com/PaddleFlow/paddle-operator/blob/main/docs/design-coschedule.md)
+
+
+## Alternatives Considered
+
+One option is to add PaddlePaddle support to the existing tf-operator, but the parameters and operations between two frameworks are quite different. Combining them may make the user experience unnecessarily complicated.
```

Comment:
This makes sense. What about the implementation of the operator? Could you leverage https://github.com/kubeflow/common's interface that's used for other operators?

Reply:
We are using kube-builder as the skeleton for the paddle-operator, part of the kubeflow common components are not necessary for our project, like job controller etc. We'll see whether we can leverage on kubeflow common for advanced features.

Reply:
Thanks for your reply, we are aware of your concern, technically, we are building our operator with the newest version of kubebuilder which take care of almost all the staff like Informer/Indexer/clienset etc. It leaves out the Reconcile function to be implemented, it also provides CRUD operations directly on resources with context, which means that the kubeflow/common may not necessary in this circumstances.